src/main/scala/abalone/variant/Variant.scala

```diff
@@ -46,128 +46,233 @@ abstract class Variant private[variant] (
-       'a' to '3' or 'c' to '4' are line moves of a single marble
-
-  we want to have :
-  1. moves of 1 marble + side moves (as we are reusing moves of 1 marble)
-  2. line moves of 2+ marbles + pushes ((re-do a pass on line moves that were stuck ?))
+       'a' to 'M' is a downRight side move of three marbles.
+       'c' to 'J' is a downLeft side move of three marbles.
+       'a' to 'L' is a downRight side move of two marbles (only 'a' and 'b' would move).
+       'c' to 'K' is a downLeft side move of two marbles (only 'c' and 'b' would move)
+       'a' to 'K' or 'c' to 'L' are line moves of a single marble
+
+  For moves of 2 marbles or more, once you get the direction of the line,
+   you can easily generate the side moves as being the one before and the one after,
+   following a rotation :
+    \   /
+   -  o  -
+    /   \
+  e.g. if you are moving upRight the side moves to consider are "upLeft" and "right".
+
+  1. moves of 1 marble
+  2. generate any possible pair of marbles and use it to generate moves of 2 and 3 marbles
   then merge these as valid moves.
   */
-  def validMoves(situation: Situation): Map[Pos, List[Move]] = {
-    val movesOf1 = validMovesOf1(situation)
-    val lineMoves = movesOf1 ++ validLineMoves(situation).map {
-      case (k, v) => k -> (v ++ movesOf1.getOrElse(k, Iterable.empty))
-    }
-    val lineMovesAndPushes = lineMoves ++ validPushes(situation).map {
-      case (k, v) => k -> (v ++ lineMoves.getOrElse(k, Iterable.empty))
-    }
-    lineMovesAndPushes ++ validSideMoves(situation).map { // @TODO: this should reuse what was computed in validMovesOf1
-      case (k, v) => k -> (v ++ lineMovesAndPushes.getOrElse(k, Iterable.empty))
-    }
-  }
+  def validMoves(situation: Situation): Map[Pos, List[Move]] =
+    (validMovesOf1(situation).toList ++ validMovesOf2And3(situation).view.mapValues(_.map(_._2)).toList)
+      .groupBy(_._1).map{case(k, v) => k -> v.map(_._2).toSeq.flatten}.toMap
 
   def validMovesOf1(situation: Situation): Map[Pos, List[Move]] =
-    situation.board.piecesOf(situation.player).flatMap {
+    turnPieces(situation).flatMap {
       case ((pos, piece)) =>
         Map(pos ->
-          pos.neighbours.flatten
-          .filterNot(situation.board.pieces.contains(_))
-          .map(landingSquare => 
-            Move(piece, pos, landingSquare, situation, boardAfter(situation, pos, landingSquare), false)
+          pos.neighboursAsDirs.flatMap(d => d(pos)).filterNot(situation.board.pieces.contains(_))
+          .map(landingSquare =>
+            Move(piece, pos, landingSquare, situation, boardAfter(situation, pos, landingSquare))
           )
         )
     }.toMap
 
-  def validSideMoves(@nowarn situation: Situation):  Map[Pos, List[Move]] = {
-    Map()
-  }
-
-  def validLineMoves(@nowarn situation: Situation):  Map[Pos, List[Move]] = {
-    Map()
-  }
-    
+  def validMovesOf2And3(situation: Situation): Map[Pos, List[(String, Move)]] = {
```

Comment:
Is the 'category' part functional? Or was it just for debugging?
As in can we rewrite to get a signature of:
`def validMovesOf2And3(situation: Situation): Map[Pos, List[(Move)]]`?